### PR TITLE
Screen Lifecycle Observation...

### PR DESCRIPTION
**Pull request by andersio**
Aims to provide a generalised mechanism that can supersede several custom BoxVC implementations currently in the codebase, and a potential mean to achieve the ultimate goal of #21 i.e. emitting analytics event only when the screen first appears.

### CODE REVIEW

File: TechnicalDocuments/Proposals/ScreenLifecycleObservation.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+`ScreenLifecycleEvent` also provides two extra properties representing two critical and unique scenarios of navigation:
+
+* `isBeingPresentedInitially`
+   If this is `true`, the screen is being presented by its container for the very first time [1]. Subsequent reappearance, e.g. user having returned from another screen in a navigation flow, would not lead to this being `true`.
+   
+    
+
+* `isBeingRemovedIndefinitely`
+   If this is `true`, the screen is about to be removed from its container **indefinitely** [2]. Temporarily disappearance, e.g. being covered by another screen in a navigation flow, would not lead to this being `true`.
```
Comment (mluisbrown) [inline TechnicalDocuments/Proposals/ScreenLifecycleObservation.md]: Change `indefinitely` to `permanently` for clarity (see previous comment).

File: TechnicalDocuments/Proposals/ScreenLifecycleObservation.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+   
+_[1] This holds until it is removed from its container indefinitely._
```
Comment (mluisbrown) [inline TechnicalDocuments/Proposals/ScreenLifecycleObservation.md]: s/indefinitely/permanently/

File: TechnicalDocuments/Proposals/ScreenLifecycleObservation.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+`didDisappear` | `viewDidDisappear`
+
+`ScreenLifecycleEvent` also provides two extra properties representing two critical and unique scenarios of navigation:
+
+* `isBeingPresentedInitially`
+   If this is `true`, the screen is being presented by its container for the very first time [1]. Subsequent reappearance, e.g. user having returned from another screen in a navigation flow, would not lead to this being `true`.
+   
+    
+
+* `isBeingRemovedIndefinitely`
```
[the remark this answers is not in the document]
Comment (sergdort) [reply TechnicalDocuments/Proposals/ScreenLifecycleObservation.md]: Isn't `viewDidLoad` not enough for these use cases?

[the remark this answers is not in the document]
Comment (mluisbrown) [reply TechnicalDocuments/Proposals/ScreenLifecycleObservation.md]: What we want here is the equivalent of `viewDidUnload`, which doesn't exist in `UIKit`.

[the remark this answers is not in the document]
Comment (sergdort) [reply TechnicalDocuments/Proposals/ScreenLifecycleObservation.md]: Oh, my bad. I mean for `isBeingPresentedInitially`

File: TechnicalDocuments/Proposals/ScreenLifecycleObservation.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+```
+
+The event model `ScreenLifecycleEvent` reflects the lifecycle status of a given screen at a given time. The status can be directly mapped to messages emitted by UIKit:
+
+Status | UIKit message
+--- | ---
+`willAppear` | `viewWillAppear`
+`didAppear` | `viewDidAppear`
+`willDisappear` | `viewWillDisappear`
+`didDisappear` | `viewDidDisappear`
```
Comment (zzcgumn) [inline TechnicalDocuments/Proposals/ScreenLifecycleObservation.md]: I think we would need to `viewDidLoad` as well. This is the best time in the view life cycle to assign the accessibility identifier for the view.

File: TechnicalDocuments/Proposals/ScreenLifecycleObservation.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+`ScreenLifecycleEvent` also provides two extra properties representing two critical and unique scenarios of navigation:
+
+* `isBeingPresentedInitially`
```
Comment (sergdort) [inline TechnicalDocuments/Proposals/ScreenLifecycleObservation.md]: `viewDidLoad` should be good enough, no?

**Comment by ilyapuchka** [comment]
I think this is a valuable proposal on its own and I support it (in fact I thought that this is something that already was done in Bento) 👍  
But I don't think that it should be used exclusively for tracking purposes and should suppress #21 , see my comment there for my reasoning.

**Comment by sergdort** [comment]
Does it mean that `BoxVM` always conforms to `ScreenLifecycleEvent ` or do we conform our VM when needed?

**Comment by andersio** [comment]
We can do it both ways: adding it as a new protocol req with default implementation, or adding a new protocol.

**Pull request merged**